### PR TITLE
Refactor subset method

### PR DESCRIPTION
**Pull request by jedwards4b**
Refactors the subset decomposition method to support multiple destinations for each source when reading.  This allows for example halo regions to be populated.    This method so far only works for the subset rearranger and the mapping file cannot be used for write since it would be in-determinant to have multiple sources for the same destination.

### CODE REVIEW

File: src/clib/pioc.c [hunk 1]
```diff
@@ -595,6 +595,10 @@ PIOc_InitDecomp(int iosysid, int pio_type, int ndims, const int *gdimlen, int ma
     /* Remember the maplen. */
     iodesc->maplen = maplen;
 
+    /* check if the decomp is valid for write or is read-only */
```
Comment (jedwards4b) [inline src/clib/pioc.c:598]: I'm not sure what to do here.   check_compmap gathers the entire compmap to comp task 0 and checks for duplicate values.  If any are found it's an error to use for write and I've added an assert to the darray_write routines to catch this.  But this is a relatively expensive and somewhat memory intense operation so my inclination is to leave it off.

**Comment by jedwards4b** [comment]
@edwardhartnett This is ready to merge, do you want to review?

File: src/clib/pio_darray.c [hunk 2]
```diff
@@ -821,8 +821,9 @@ PIOc_write_darray(int ncid, int varid, int ioid, PIO_Offset arraylen, void *arra
     bufptr = (void *)((char *)wmb->data + arraylen * iodesc->mpitype_size * wmb->num_arrays);
     if (arraylen > 0)
     {
+        PLOG((3, "copying %ld bytes of user data %d", arraylen * iodesc->mpitype_size, iodesc->mpitype_size));
+        printf(( "copying %ld bytes of user data %d num_arrays %d bufptr %x array %x\n", arraylen * iodesc->mpitype_size, iodesc->mpitype_size, wmb->num_arrays, bufptr, array));
```
Comment (edwardhartnett) [inline src/clib/pio_darray.c]: I'm sure you don't want to leave printf statements in...

Is the logging not working for you? Build with --enable-logging and you can add log statements without using printf...

**Comment by edwardhartnett** [comment]
@jedwards4b looks good. I will let you merge.

After this is merged, are you ready for release?

File: src/clib/pio_darray.c (hunk 2 above)
Comment (jedwards4b) [reply src/clib/pio_darray.c]: I'm confused, this line was already removed.  https://github.com/jedwards4b/ParallelIO/blob/refactor_subset_method/src/clib/pio_darray.c#L825

Comment (jedwards4b) [reply src/clib/pio_darray.c]: logging works but sometimes while debugging I find it easier to enable a single print instead of all of the level 3 prints

**Comment by jedwards4b** [comment]
@edwardhartnett Yes, ready for a release.  Thanks

**Pull request merged**